### ISSUE TITLE
Credit for submitting incorrect work

### ISSUE THREAD

**Issue by dsanson** [Carnap/Carnap#154]
In certain contexts, I'd like to be able to give students credit for submitting work, even if it is incorrect. For example, in my "textbook", I have included open-ended short-answer questions, and I'd like to give any students who *submit* something full credit, no matter what it is. And I'd like to offer the same for other types of exercises---for example, when introducing derivations, I'd like to encourage them to stop and *try* a derivation, and give them credit for *trying*, even if they can't get it right. For this, I can put the derivation in "exam" mode, and allow incorrect submissions, and then give anyone who submits anything full credit.

At the moment, I can see on Carnap *that* they submitted something. And I can go through and manually apply credit to all those submissions. But what I'd really like is a toggle that just gives them full credit no matter what.

An alternative workaround would be to expose the submission data to the CSV export. That way, I could give them credit on my spreadsheet, even if they don't get credit on Carnap.

**Comment by gleachkr** [comment]
I was thinking about this more, and wondering what you might think about the following approach. We could add an `give-credit` option to short answer questions. We keep the current behavior if this is missing, but we introduce `give-credit="onSubmission"` to automatically give credit, and later add `give-credit="onKeywords:...."` to handle #155. We've already got a way of making sure that any answer is correct for mulitple choice, there's an issue #160 that would add the same for multiple selections.

So, then for the use case you're thinking about, the approach would be to couple the "just try" exercise (derivation, truth table...) with an auto-credited short answer qualitative exercise, where the student does a little metacognition on what's giving them trouble, or a multiple-choice/selection where they in some other way offer feedback to the instructor. The just try exercise could be for any number of points, including zero, or you could chain disjunctively, giving, e.g. 3 points for completion or 2 points for explaining what's giving you trouble.

What do you think? Any downsides to this?

**Comment by dsanson** [comment]
That sounds interesting. Auto-credited short answer qualitative exercises would be useful no matter what. They can be used to ask students to engage in metacognition. They can also be used in low stakes assignments to ask students to do all sorts of other open-ended exercises.

A simple "I tried" question can be a multiple-choice question behind the scenes, but would best be presented as a mono-choice question, so, essentially just a button, for frictionless one-click completion and submission. Mono-choice questions like this seem like they could be useful in lots of unanticipated ways.

For something like a derivation, I'd prefer it if the "I tried" button was *presented* as an integrated part of the same exercise. I worry that if students see it as a separate thing, then (a) some of them will fail to do it and (b) it will encourage some of them to just click through without actually trying. But again, that seems like it is mostly about presentation. Behind the scenes, Carnap could conceptualize it as two exercises: a derivation with submission=none and a mono-choice question.

What I'm imagining, from a UI perspective, is a derivation that has its submit button replaced or supplemented by an "I tried" button. My original idea was that I could get this effect by putting the problem into exam mode, and having it set to get full points whether or not it is right or wrong. But I like your idea better, and I like the idea of having *both* buttons, even if I plan to give full points either way. That way they could still benefit from all the error checking that Carnap provides. If they get the problem right, they get the automated kudos. If they can't get it right, they encounter resistance and it is clearly communicated that they don't quite get it. But they have an easy bailout option. 

Some sort of disjunctive logic between the two buttons seems appropriate. And the option to give more points for getting it right. But the simplest version of that isn't great: a student who clicks "I tried" should be able to go back and successfully "submit", if they figure out the solution before the due date.

**Comment by gleachkr** [comment]
OK, I've gone ahead and added the `give-credit` option. I am thinking about how best to implement the "I tried" button. I wonder if this might be a job for a javascript extension. The extension could

0. look for pairs of exercises with the same label that are enclosed together in a div of a certain kind.
1. add an "I tried" button to the list of buttons associated with the first exercise
2. set the second exercise to have `display:none`
2. for onclick of the button, set the first exercise to have `display:none`, and set the second exercise to have `display:block`

So basically, this would be one form of problem-chaining, where you can bail out on one problem and opt to complete another instead.

I think this also would work best in combination with some version of #129, to handle the last issue you mention (that a student figures things out later, but still before the due date).
